src/coreComponents/common/LifoStorageCommon.hpp

```diff
@@ -0,0 +1,295 @@
+ */
+template< typename T, typename INDEX_TYPE >
+class LifoStorageCommon
+{
+protected:
+
+  /// number of buffers to be inserted into the LIFO
+  int m_maxNumberOfBuffers;
+  /// size of one buffer in bytes
+  size_t m_bufferSize;
+  /// name used to store data on disk
+  std::string m_name;
+  /// Queue of data stored on host memory
+  FixedSizeDequeWithMutexes< T, INDEX_TYPE > m_hostDeque;
+
+  /// counter of buffer stored in LIFO
+  int m_bufferCount;
+  /// counter of buffer pushed to host
+  int m_bufferToHostCount;
+  /// counter of buffer pushed to disk
+  int m_bufferToDiskCount;
+
+
+  /// condition used to tell m_worker queue has been filled or processed is stopped.
+  std::condition_variable m_task_queue_not_empty_cond[2];
+  /// mutex to protect access to m_task_queue.
+  std::mutex m_task_queue_mutex[2];
+  /// queue of task to be executed by m_worker.
+  std::deque< std::packaged_task< void() > > m_task_queue[2];
+  /// thread to execute tasks.
+  std::thread m_worker[2];
+  /// boolean to keep m_worker alive.
+  bool m_continue;
+  /// marker to detect first pop
+  bool m_hasPoppedBefore;
+public:
+
+
+  /**
+   * A LIFO storage will store numberOfBuffersToStoreDevice buffer on
+   * deevice, numberOfBuffersToStoreHost on host and the rest on disk.
+   *
+   * @param name                           Prefix of the files used to save the occurenncy of the saved buffer on disk.
+   * @param elemCnt                        Number of elments in the LvArray we want to store in the LIFO storage.
+   * @param numberOfBuffersToStoreOnHost   Maximum number of array to store on host memory ( -1 = use 80% of remaining memory ).
+   * @param maxNumberOfBuffers             Number of arrays expected to be stores in the LIFO.
+   */
+  LifoStorageCommon( std::string name, size_t elemCnt, int numberOfBuffersToStoreOnHost, int maxNumberOfBuffers ):
+    m_maxNumberOfBuffers( maxNumberOfBuffers ),
+    m_bufferSize( elemCnt*sizeof( T ) ),
+    m_name( name ),
+    m_hostDeque( numberOfBuffersToStoreOnHost, elemCnt, LvArray::MemorySpace::host ),
+    m_bufferCount( 0 ), m_bufferToHostCount( 0 ), m_bufferToDiskCount( 0 ),
+    m_continue( true ),
+    m_hasPoppedBefore( false )
+  {
+    m_worker[0] = std::thread( &LifoStorageCommon< T, INDEX_TYPE >::wait_and_consume_tasks, this, 0 );
+    m_worker[1] = std::thread( &LifoStorageCommon< T, INDEX_TYPE >::wait_and_consume_tasks, this, 1 );
+  }
+
+  virtual ~LifoStorageCommon()
+  {
+    m_continue = false;
+    m_task_queue_not_empty_cond[0].notify_all();
+    m_task_queue_not_empty_cond[1].notify_all();
+    m_worker[0].join();
+    m_worker[1].join();
+  }
+
+  /**
+   * Asynchroneously push a copy of the given LvArray into the LIFO
+   *
+   * @param array The LvArray to store in the LIFO, should match the size of the data used in constructor.
+   */
+  virtual void pushAsync( arrayView1d< T > array ) = 0;
+
+  /**
+   * Waits for last push to be terminated
+   */
+  virtual void pushWait() = 0;
```

Comment:
Do you have to define `virtual` members in this `class`?

Reply:
@TotoGaz any advice on that ?